### ISSUE TITLE
Results of query randomly ordered; `ORDER BY` not working

### ISSUE THREAD

**Issue by sploiselle** [cockroachdb/cockroach#12155]
### Steps to Repro

1. Download [gitter-app-schema.zip](https://github.com/cockroachdb/cockroach/files/637797/gitter-app-schema.zip) and unzip.
2. With Cockroach (version beta-20161201) running, run the following commands:
  ```
  cockroach sql --execute="CREATE DATABASE gitter;"
  cockroach sql --database=gitter < gitter_messages.sql
  cockroach sql --database=gitter < gitter_employees.sql
  cockroach sql --database=gitter < gitter_visitors.sql
  cockroach sql
  ```
3. Run the following query:
  ```
  SELECT sent_by, COUNT(*) AS count FROM gitter.messages WHERE (sent_by IN (SELECT id FROM visitors)) AND (time_sent BETWEEN TIMESTAMP WITH TIME ZONE '2016-10-08' AND TIMESTAMP WITH TIME ZONE '2016-12-07') GROUP BY sent_by ORDER BY count DESC;
  ```

### Expected Results

Result table is ordered by `count` column.

### Actual Result

Table is not ordered whatsoever. Running the query also produces different results every time.

**Comment by tamird** [comment]
FYI this query can be simplified to
```sql
SELECT sent_by, COUNT(*) FROM gitter.messages INNER JOIN gitter.visitors ON messages.sent_by = visitors.id WHERE time_sent BETWEEN '2016-10-08' AND '2016-12-07' GROUP BY sent_by ORDER BY COUNT(*) DESC;
```
But indeed, ordering by the aggregate is broken. Ordering by `sent_by` does work.

**Comment by sploiselle** [comment]
Thanks for the query optimization, @tamird

**Comment by knz** [comment]
Many thanks for spotting this Sean :)

**Comment by sploiselle** [comment]
@knz Hate to circle back on this, but I'm still experiencing the same issue with my first (poorly written) query.

```
SELECT sent_by, COUNT(*) AS count FROM gitter.messages WHERE sent_by IN (SELECT id FROM gitter.visitors) AND time_sent BETWEEN TIMESTAMPTZ '2016-10-08' AND '2016-12-07' GROUP BY sent_by ORDER BY count DESC LIMIT 5;
```

The results I get back from this have a sporadic ordering.

However, if I used Tamir's improved query, the ordering returns consistently.

```
SELECT sent_by, COUNT(*) FROM gitter.messages INNER JOIN gitter.visitors ON messages.sent_by = visitors.id WHERE time_sent BETWEEN '2016-10-08' AND '2016-12-07' GROUP BY sent_by ORDER BY COUNT(*) DESC LIMIT 5;
```

I'm able to make forward progress thanks to Tamir rewriting the query, but seems like there's possibly still a bug present.

**Comment by knz** [comment]
Yes that's a bug :-)

Don't hate circling back. Grit is commendable! 

Maybe related to #12525.

**Comment by knz** [comment]
I believe this has been fixed by #12548. At least with the queries given here I now observe a consistent sort order.

@sploiselle can you check whether the behavior is now OK?

**Comment by knz** [comment]
I've been testing this extensively and cannot reproduce any more. I believe that issue has been fixed satisfactorily. Feel free to re-open if there's a new repro.
